File: bracket_algorithm.cs
// C# program for the above approach
using System;
using System.Collections.Generic;

public class GFG{
class pair : IComparable<pair>
	{
		public int first,second;
		public pair(int first, int second)
		{
			this.first = first;
			this.second = second;
		}

		public int CompareTo(pair p)
		{
			return this.first - p.first;
		}
	}

// Function to check possible RBS from
// the given Strings
static int checkRBS(String[] S)
{
	int N = S.Length;

	// Stores the values {sum, min_prefix}
	pair []v = new pair[N];

	// Iterate over the range
	for (int i = 0; i < N; ++i) {
		String s = S[i];

		// Stores the total sum
		int sum = 0;

		// Stores the minimum prefix
		int pre = 0;
		foreach (char c in s.ToCharArray()) {
			if (c == '(') {
				++sum;
			}
			else {
				--sum;
			}

			// Check for minimum prefix
			pre = Math.Min(sum, pre);
		}

		// Store these values in vector
		v[i] = new pair( sum, pre );
	}

	// Make two pair vectors pos and neg
	List<pair> pos = new List<pair>();
	List<pair > neg = new List<pair>();

	// Store values according to the
	// mentioned approach
	for (int i = 0; i < N; ++i) {
		if (v[i].first >= 0) {
			pos.Add(
				new pair( -v[i].second, v[i].first ));
		}
		else {
			neg.Add(
					new pair( v[i].first - v[i].second,
				-v[i].first ));
		}
	}

	// Sort the positive vector
	pos.Sort();

	// Stores the extra count of
	// open brackets
	int open = 0;
	foreach (pair p in pos) {
		if (open - p.first >= 0) {
			open += p.second;
		}

		// No valid bracket sequence
		// can be formed
		else {
			Console.Write("No"
				+ "\n");
			return 0;
		}
	}

	// Sort the negative vector
	neg.Sort();

	// Stores the count of the
	// negative elements
	int negative = 0;
	foreach (pair p in neg) {

		if (negative - p.first >= 0) {
			negative += p.second;
		}

		// No valid bracket sequence
		// can be formed
		else {
			Console.Write("No\n");
			return 0;
		}
	}

	// Check if open is equal to negative
	if (open != negative) {
		Console.Write("No\n");
		return 0;
	
	}
	Console.Write("Yes\n");
	return 0;
}

// Driver Code
public static void Main(String[] args)
{
	String []arr = { ")", "()(" };
	checkRBS(arr);

}
}

// This code is contributed by 29AjayKumar

File: bracket_algorithm.py
# Python 3 program for the above approach

# Function to check possible RBS from
# the given strings
def checkRBS(S):

	N = len(S)

	# Stores the values {sum, min_prefix}
	v = [0]*(N)

	# Iterate over the range
	for i in range(N):
		s = S[i]

		# Stores the total sum
		sum = 0

		# Stores the minimum prefix
		pre = 0
		for c in s:
			if (c == '('):
				sum += 1

			else:
				sum -= 1

			# Check for minimum prefix
			pre = min(sum, pre)

		# Store these values in vector
		v[i] = [sum, pre]

	pos = []
	neg = []

	# Store values according to the
	# mentioned approach
	for i in range(N):
		if (v[i][0] >= 0):
			pos.append(
				[-v[i][1], v[i][0]])

		else:
			neg.append(
				[v[i][0] - v[i][1],
				-v[i][0]])

	# Sort the positive vector
	pos.sort()

	# Stores the extra count of
	# open brackets
	open = 0
	for p in pos:
		if (open - p[0] >= 0):
			open += p[1]

		# No valid bracket sequence
		# can be formed
		else:
			print("No")

			return 0

	# Sort the negative vector
	neg.sort()

	# Stores the count of the
	# negative elements
	negative = 0
	for p in neg:

		if (negative - p[0] >= 0):
			negative += p[1]

		# No valid bracket sequence
		# can be formed
		else:
			print("No")
			return 0

	# Check if open is equal to negative
	if (open != negative):
		print("No")
		return 0

	print("Yes")

# Driver Code
if __name__ == "__main__":

	arr = [")", "()("]
	checkRBS(arr)

	# This code is contributed by ukasp.
